python/ray/data/_internal/execution/interfaces/op_runtime_metrics.py

```diff
@@ -485,6 +509,12 @@ def on_task_output_generated(self, task_index: int, output: RefBundle):
         for block_ref, meta in output.blocks:
             assert meta.exec_stats and meta.exec_stats.wall_time_s
             self.block_generation_time += meta.exec_stats.wall_time_s
+
+            # Should these go into a different callback (perhaps task_finished?)
+            self.task_cpu_time += meta.exec_stats.cpu_time_s
+            self.task_udf_time += meta.exec_stats.udf_time_s
+            self.task_idle_time += meta.exec_stats.wall_time_s - meta.exec_stats.cpu_time_s # how does UDF time fit into this?
```

Comment:
i think this is the right place for `task_cpu_time` and `task_udf_time`. if i understand your intent correctly, for `task_idle_time`, i think we need some additional information. we need to get the difference between (time at which current task finished) and (time at which next task starts)

Reply:
I think what I was aiming for here is how long we are idle _within_ a task, do you think that is possible here? Some situations I would hope to measure with it would be time spent between `yields` of a generator task, or from streaming backpressure due to hitting the max number of streaming blocks. Is that possible here?

Reply:
> time spent between yields of a generator task
Would this be the same as the time it takes for the generator to produce the next output? or more like the timing around here when we yield batches from the iterator? https://github.com/ray-project/ray/blob/master/python/ray/data/iterator.py#L185-L186

> from streaming backpressure due to hitting the max number of streaming blocks
we have this existing metric `task_submission_backpressure_time`, which keeps track of the cumulative time spent in backpressure to submit the next task. is this what you had in mind for your use case?
https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/execution/interfaces/op_runtime_metrics.py#L204-L210

Reply:
More like the timing around when we yield batches from the iterator. 

As for the `task_submission_backpressure_time`, I'm still a bit confused on how to interpret metric. It seems like this metric will increase as long as we have enough tasks submitted to our pool of workers. But that would always seem like the case given that the scheduling loop runs every millisecond or so. For a one-operator pipeline, what does an increasing or stagnated `task_submission_backpressure_time` indicate? I don't think it would mean that we're necessarily producing blocks faster than consuming them, because the running tasks themselves could just be slow, right?

I think Raul kind of mentions in this [comment](https://github.com/ray-project/ray/pull/43110#discussion_r1503323275) what I'd like to capture. When do we apply backpressure to running tasks? How can we measure that?

Reply:
> For a one-operator pipeline, what does an increasing or stagnated task_submission_backpressure_time indicate? I don't think it would mean that we're necessarily producing blocks faster than consuming them, because the running tasks themselves could just be slow, right?

Yeah, the reason could be several possible things, but the underlying cause is that there are too many tasks queued for the operator, so we don't want to add any more.

[This](https://github.com/ray-project/ray/blob/33ee73266060ab80db67f04afc45cc8849f81bbf/python/ray/data/_internal/execution/streaming_executor_state.py#L532-L534) is where we evaluate whether the op should be backpressured. I think older Ray versions may look slightly different, but essentially `backpressure_policy.can_add_input()` returns False if backpressure is to be applied. When backpressure is applied, the operator will no longer accept new inputs.